### ISSUE TITLE
Adding 'Order by' or 'Group by' keywords to query results with Error

### ISSUE THREAD

**Issue by patryk-plura** [turbot/steampipe-plugin-azuread#109]
**Describe the bug**
Adding 'order by' or 'group by' keywords to query ends with:
```
Error: rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime (SQLSTATE HV000)
```

**Steampipe version (`steampipe -v`)**
v0.19.5

**Plugin version (`steampipe plugin list`)**
0.9.0

**To reproduce**
Working SQL:
```
select
        targetResources ->> 'id' as TargetId,
        auditLogs.activity_date_time as LogTimestamp
from
        azuread_directory_audit_report as auditLogs,
        jsonb_array_elements(auditLogs.additional_Details) as additionalDetails,
        jsonb_array_elements(auditLogs.target_resources) as targetResources
where
        auditLogs.activity_display_name = 'Exchange token'
        and auditLogs.result = 'success'
limit 10
```
Output:
```
+--------------------------------------+---------------------------+
| targetid                             | logtimestamp              |
+--------------------------------------+---------------------------+
| 255ffbbc-4514-4499-9132-4bac8727b314 | 2023-05-08T12:18:54+02:00 |
| 255ffbbc-4514-4499-9132-4bac8727b314 | 2023-05-08T12:18:54+02:00 |
| 255ffbbc-4514-4499-9132-4bac8727b314 | 2023-05-08T12:18:54+02:00 |
| 255ffbbc-4514-4499-9132-4bac8727b314 | 2023-05-08T12:18:54+02:00 |
| 255ffbbc-4514-4499-9132-4bac8727b314 | 2023-05-08T12:18:54+02:00 |
| 255ffbbc-4514-4499-9132-4bac8727b314 | 2023-05-08T12:18:54+02:00 |
| 255ffbbc-4514-4499-9132-4bac8727b314 | 2023-05-08T12:18:54+02:00 |
| 255ffbbc-4514-4499-9132-4bac8727b314 | 2023-05-08T12:18:54+02:00 |
| 9696f784-1819-4bc9-9325-5ac2206b4682 | 2023-05-08T12:18:44+02:00 |
| 9696f784-1819-4bc9-9325-5ac2206b4682 | 2023-05-08T12:18:44+02:00 |
+--------------------------------------+---------------------------+
```

Broken SQL:
```
select
        targetResources ->> 'id' as TargetId,
        auditLogs.activity_date_time as LogTimestamp
from
        azuread_directory_audit_report as auditLogs,
        jsonb_array_elements(auditLogs.additional_Details) as additionalDetails,
        jsonb_array_elements(auditLogs.target_resources) as targetResources
where
        auditLogs.activity_display_name = 'Exchange token'
        and auditLogs.result = 'success'
order by
        LogTimestamp
limit 10
```
Output:
```
Error: rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime (SQLSTATE HV000)
```

**Expected behavior**
Adding those keywords shouldn't end with error.

**Additional context**
I use client secret authentication (copy of azuread.spc):
```
connection "azuread" {
  plugin = "azuread"
  tenant_id = "guid"
  client_id = "guid"
  client_secret = "secret"
}
```

Added scopes to the application in AAD:
- AuditLog.Read.All
- Directory.Read.All
- User.Read.All

But something like that works perfectly:
```
select 
	id, 
	title 
from 
	azuread_user 
order by 
	id 
limit 10
```

Output from logs (database.log):
```
2023-05-08 10:41:46.509 UTC [WARN]  hub: stream receive error rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime (0xc000212d80)
2023-05-08 10:41:46.509 UTC [2272] ERROR:  rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime
2023-05-08 10:41:46.509 UTC [2272] STATEMENT:  select
	        targetResources ->> 'id' as TargetId,
	        auditLogs.activity_date_time as LogTimestamp
	from
	        azuread_directory_audit_report as auditLogs,
	        jsonb_array_elements(auditLogs.additional_Details) as additionalDetails,
	        jsonb_array_elements(auditLogs.target_resources) as targetResources
	where
	        auditLogs.activity_display_name = 'Exchange token'
	        and auditLogs.result = 'success'
	order by
	        LogTimestamp
	limit 10
2023-05-08 10:41:48.762 UTC [2270] LOG:  could not receive data from client: Connection reset by peer
2023-05-08 10:41:48.762 UTC [2270] LOG:  disconnection: session time: 0:01:59.995 user=root database=steampipe host=127.0.0.1 port=34206
2023-05-08 10:41:48.762 UTC [2269] LOG:  could not receive data from client: Connection reset by peer
2023-05-08 10:41:48.762 UTC [2269] LOG:  disconnection: session time: 0:02:00.010 user=root database=steampipe host=127.0.0.1 port=34192

```

Output from logs (plugin.log)
```
2023-05-08 10:41:33.615 UTC [WARN]  steampipe-plugin-azuread.plugin: [WARN]  1683542473956: waitForPendingItem timed out waiting for pending transfer (azuread-1683542473956) 0xc016c5cc90 indexBucketKey: index__azuread_azuread_directory_audit_report, item key azuread_azuread_directory_audit_report_--activity_display_name-=-Exchange token---activity_display_name-=-Exchange token--activity_display_name-=-Exchange token-result-=-success-activity_display_name-=-Exchange token-result-=-success_activity_date_time,additional_details,target_resources,activity_display_name,result_-1
2023-05-08 10:41:46.508 UTC [WARN]  steampipe-plugin-azuread.plugin: [WARN]  1683542473956: recovered a panic from a wrapped hydrate function: interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime
2023-05-08 10:41:46.508 UTC [WARN]  steampipe-plugin-azuread.plugin: [WARN]  1683542473956: stack: goroutine 91245 [running]:
2023-05-08 10:41:46.508 UTC [WARN]  steampipe-plugin-azuread.plugin: [WARN]  1683542473956: doList callHydrateWithRetries (azuread-1683542473956) returned err rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime
2023-05-08 10:41:46.508 UTC [WARN]  steampipe-plugin-azuread.plugin: [WARN]  1683542473956: QueryData StreamError rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime (azuread-1683542473956)
2023-05-08 10:41:46.508 UTC [WARN]  steampipe-plugin-azuread.plugin: [WARN]  1683542473956: streamRows for azuread-1683542473956 - execution has failed (rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime) - calling queryCache.AbortSet
2023-05-08 10:41:46.508 UTC [WARN]  steampipe-plugin-azuread.plugin: [WARN]  1683542473956: executeForConnection azuread returned error rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime
2023-05-08 10:41:46.508 UTC [WARN]  steampipe-plugin-azuread.plugin: [WARN]  1683542473956: error channel received rpc error: code = Internal desc = hydrate function listAdDirectoryAuditReports failed with panic interface conversion: *models.SignIn is not models.DirectoryAuditable: missing method GetActivityDateTime

```

I've tried same query but on other AAD tenant (same AAD application scopes) and works perfectly.

**Comment by misraved** [comment]
Apologies for the bump @patryk-plura and thanks for raising this issue with such comprehensive details 👍.

Just for clarification, when you say `I've tried the same query but on other AAD tenant (same AAD application scopes) and works perfectly.`, do you mean that you only see this issue with a particular tenant and not others?

**Comment by patryk-plura** [comment]
Any update on this?

**Comment by rajlearner17** [comment]
@patryk-plura, Thanks for reaching out. Apology for staying silent on it. 
We are checking and will update you shortly.

@ParthaI

**Comment by ParthaI** [comment]
Hi @patryk-plura,
I have tried the provided query with activity_display_name like (`Update service principal`, `Remove service principal credentials`,  `Add service principal` and `Hard delete service principal`) rather than 'Exchange token'. Because we do not have any rows with activity_display_name 'Exchange token'.

```sql
select
  targetResources ->> 'id' as TargetId,
  auditLogs.activity_date_time as LogTimestamp
from
  azuread_directory_audit_report as auditLogs,
  jsonb_array_elements(auditLogs.additional_Details) as additionalDetails,
  jsonb_array_elements(auditLogs.target_resources) as targetResources
where
  auditLogs.activity_display_name = 'Update service principal'
  and auditLogs.result = 'success'
order by
  LogTimestamp
limit 10;
```

This query works with two different tenants, with & without a limit clause. However, could not replicate the issue.
We have a limited amount of data (Near about 50 rows) in our tenants.
Additionally, here is the code block where we are converting the result to type [DirectoryAudit](https://github.com/turbot/steampipe-plugin-azuread/blob/main/azuread/table_azuread_directory_audit_report.go#L151)
 
I suspect this one results from an event type we don’t expect from the API/SDK since using order by or group by will result in getting all results and not using any quals passed in.
The API which is being used here only returns the result of the type [DirectoryAuditable](https://github.com/microsoftgraph/msgraph-sdk-go/blob/main/models/directory_audit.go#L428) , not any Sign_In report type.

**Here are a few options to test**
1- Please try with auditLogs.activity_display_name = 'Update service principal' or any other activity name to ensure it renders the same issue for you. Again this is for testing purposes only.
2- We have made a minor update in the branch `issue-109` if you can checkout and try it.

Steps for checkout

- git clone https://github.com/turbot/steampipe-plugin-azuread.git
- cd steampipe-plugin-azuread
- git checkout `issue-109`
- make
- Run steampipe query


Sorry that we cannot reproduce this; your input will help us pin down the issue. Thank you!

**Comment by cbruno10** [comment]
@ParthaI In the SDK, are you able to find types the API we're using to list events could return? I saw your link to the `DirectoryAuditable` model, but curious if you found anything in the function definition.

Also, in our code, we could handle sign in event types as well and stream those? Are the fields different than the type we're handling today?

**Comment by ParthaI** [comment]
@cbruno10, Here is the [function definition](https://github.com/microsoftgraph/msgraph-sdk-go/blob/main/auditlogs/directory_audits_directory_audit_item_request_builder.go#L77) which is used to get the details of Directory Audit Report. 

---

> Also, in our code, could we handle sign-in event types as well and stream those? Are the fields different than the type we're handling today?

We have two separate tables in the Azure-AD plugin `azuread_directory_audit_report` and `azuread_sign_in_report`, both of the table structure and the APIs response are different. [DirectoryAuditable](https://github.com/microsoftgraph/msgraph-sdk-go/blob/main/models/directory_audit.go#L428), [SignInable Model](https://github.com/microsoftgraph/msgraph-sdk-go/blob/main/models/sign_in.go#L840) 

---

> I suppose a larger question is even if we're able to stream those, would those be a part of this table or a new table (the fields in sign-in events would help us determine that).
 
Since the SignIn API returns different [type of response](https://github.com/microsoftgraph/msgraph-sdk-go/blob/main/models/sign_in.go#L840) with a different set of columns, we need a separate table for it. We already have a table named `azuread_sign_in_report` to capture those data.

**Comment by cbruno10** [comment]
@ParthaI Thanks for the info, I had missed we already had a table for these types of events!

Can you please raise a PR for the changes you've made?

Also, do we need to update any other tables that use the `client.AuditLogs().DirectoryAudits().Get(ctx, options)` call that need to filter out types like the `azuread_directory_audit_report` table does?

**Comment by ParthaI** [comment]
@cbruno10, I have submitted a [pull request](https://github.com/turbot/steampipe-plugin-azuread/pull/116) for the modifications in the `issue-109` branch. Kindly review the changes and provide your feedback on their feasibility. Thank you!

There are two tables, namely `azuread_directory_audit_report` and `azuread_sign_in_report`, which are using the `AuditLogs` API to retrieve and display their results. To ensure data consistency, I have incorporated a type check on the response before performing the `StreamList` operation. While no issues have been reported for the `azuread_sign_in_report` table in this regard, I deemed it prudent to include this check as a preventive measure.
